src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -607,60 +609,93 @@ bool Thread::IsHijackTarget(void * address)
     return false;
 }
 
-bool Thread::Hijack()
+void Thread::Hijack()
 {
     ASSERT(ThreadStore::GetCurrentThread() == ThreadStore::GetSuspendingThread());
-
     ASSERT_MSG(ThreadStore::GetSuspendingThread() != this, "You may not hijack a thread from itself.");
 
     if (m_hPalThread == INVALID_HANDLE_VALUE)
     {
         // cannot proceed
-        return false;
+        return;
     }
 
-    // requires THREAD_SUSPEND_RESUME / THREAD_GET_CONTEXT / THREAD_SET_CONTEXT permissions
+#if defined(TARGET_ARM64) && defined(TARGET_UNIX)
+    // TODO: RhpGcProbe and related asm helpers NYI for ARM64/UNIX.
+    //       disabling hijacking for now.
+    return;
+#endif
 
-    Thread* pCurrentThread = ThreadStore::GetCurrentThread();
-    uint32_t result = PalHijack(m_hPalThread, HijackCallback, this);
-    return result == 0;
+    // PalHijack will call HijackCallback or make the target thread call it.
+    // It may also do nothing if the target thread is in inconvenient state.
+    PalHijack(m_hPalThread, this);
 }
 
-UInt32_BOOL Thread::HijackCallback(HANDLE /*hThread*/, PAL_LIMITED_CONTEXT* pThreadContext, void* pCallbackContext)
+void Thread::HijackCallback(NATIVE_CONTEXT* pThreadContext, void* pThreadToHijack)
 {
-    Thread* pThread = (Thread*) pCallbackContext;
+    Thread* pThread = (Thread*) pThreadToHijack;
+    if (pThread == NULL)
+    {
+        pThread = ThreadStore::GetCurrentThread();
+
+        ASSERT(pThread != NULL);
+        ASSERT(pThread != ThreadStore::GetSuspendingThread());
+    }
 
     // we have a thread stopped, and we do not know where exactly.
     // it could be in a system call or in our own runtime holding locks
     // current thread should not block or allocate while we determine whether the location is in managed code.
-    if (pThread->CacheTransitionFrameForSuspend())
+
+    if (pThread->m_pTransitionFrame != NULL)
     {
         // This thread has already made it to preemptive (posted a transition frame)
         // we do not need to hijack it
-        return true;
+        return;
     }
 
-    void* pvAddress = (void*)pThreadContext->IP;
+    void* pvAddress = (void*)pThreadContext->GetIp();
     RuntimeInstance* runtime = GetRuntimeInstance();
     if (!runtime->IsManaged(pvAddress))
     {
         // Running in cooperative mode, but not managed.
         // We cannot continue.
-        return false;
+        return;
+    }
+
+    if (pThread->IsDoNotTriggerGcSet())
+    {
+        return;
     }
 
     ICodeManager* codeManager = runtime->GetCodeManagerForAddress(pvAddress);
```

Comment:
Do we need to check here that the thread still wants to be hijacked? The hijack signal can be delivered very late - once the GC run and everything is running again. I do not think we want to hijack the thread in that case.

Reply:
We can check suspension trap flag and do nothing if it is not set. There is no strong contract on hijack to work or not, so even if there are races, the outcome would be legal.

When I thought about this, I thought that we should optimize for common case. - If we often find that the signal is late, we should check, if the check will be typically positive (i.e. we should hijack), then we do not need to check.

My feel was that late arriving signals would be uncommon. At least on Linux, where we use realtime signals.

Reply:
We can gather some statistics here although I think it should happen after tuning the loop that sends signals. It seems too aggressive right now. I see threads interrupted in the same location over and over. IMO, that is a bigger problem than occasional redundant hijack.

Tuning the aggressiveness of the loop might impact the stats of how signals arrive, so it could be too early to measure that.

Reply:
HijackCallback might be a better place for such check - none of the work needs to happen if we are no longer suspending.

Reply:
Ah right, this _is_ in the `HijackCallback` :-)

Reply:
On the other hand checking the trap flag would be very cheap, so I think I will just add it.